code/datums/components/transfers.dm

```diff
@@ -0,0 +1,227 @@
+
+#define cant_do_shit(nerd) (!isliving(nerd) || isintangible(nerd) || is_incapacitated(nerd))
+
+/// Provides ability to output items directly into item transfer-supporting things.
+/datum/component/transfer_output
+	dupe_mode = COMPONENT_DUPE_UNIQUE_PASSARGS
+	var/atom/output_target
+
+/datum/component/transfer_output/Initialize()
+	if (!isatom(parent))
+		return COMPONENT_INCOMPATIBLE
+
+/datum/component/transfer_output/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_ATOM_MOUSEDROP, .proc/handle_drop)
+	RegisterSignal(parent, COMSIG_OUTGOING_TRANSFER, .proc/handle_outgoing)
+
+/datum/component/transfer_output/proc/handle_outgoing(comsig_target, obj/item/outgoing)
+	if(!output_target)
+		return
+
+	if(!in_interact_range(parent, output_target))
+		src.output_target = null
+		return
+
+	if(isturf(output_target))
+		outgoing.set_loc(output_target)
+		return TRUE
+
+	return SEND_SIGNAL(output_target, COMSIG_INCOMING_TRANSFER, outgoing)
+
+/datum/component/transfer_output/proc/handle_drop(comsig_target, mob/dropper, atom/over_object)
+	if(cant_do_shit(dropper))
+		return
+
+	if(over_object == parent)
+		boutput(dropper, "<span class='notice'>You reset the output location of [parent]!</span>")
+		src.output_target = null
+		return
+
+	if(!in_interact_range(parent, dropper))
+		return
+
+	if(!in_interact_range(over_object, parent))
+		boutput(dropper, "<span class='alert'>[over_object] is too far away!</span>")
+		return
+
+	if (isturf(over_object) || SEND_SIGNAL(over_object, COMSIG_CAN_LINK, parent))
+		boutput(dropper, "<span class='notice'>You set [parent] to output to [over_object].</span>")
+		output_target = over_object
+	else
+		boutput(dropper, "<span class='alert'>\The [over_object] cannot be used as an output for [parent].</span>")
+
+
+/// Provides many common item input features.
+/datum/component/transfer_input
+	dupe_mode = COMPONENT_DUPE_UNIQUE_PASSARGS
+	/// Optional proc name on parent, for transfer side-effects.
+	var/transfer_proc
+	var/list/filter
+	var/filter_proc
+	var/filter_link_proc
+
+/datum/component/transfer_input/Initialize(list/filter=null, transfer=null, filter_proc=null, filter_link_proc=null)
+	if(!isatom(parent))
+		return COMPONENT_INCOMPATIBLE
+	src.transfer_proc = transfer
+	src.filter = filter
+	src.filter_proc = filter_proc
+	src.filter_link_proc = filter_link_proc
+
+/datum/component/transfer_input/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_INCOMING_TRANSFER, .proc/handle_incoming)
+	RegisterSignal(parent, COMSIG_CAN_LINK, .proc/handle_incoming_link)
+	RegisterSignal(parent, COMSIG_ATTACKBY, .proc/handle_attackby)
+
+/datum/component/transfer_input/proc/handle_incoming_link(comsig_target, obj/other)
+	return !filter_link_proc || call(parent, filter_link_proc)(other)
+
+/datum/component/transfer_input/proc/handle_incoming(comsig_target, obj/item/incoming)
+	if (!isitem(incoming))
+		return
+
+	if (filter && !(incoming.type in filter))
```

Comment:
Conceptually I don't like stuff like this because it doesn't handle subtypes (unless in `Initialize` the `filter` var always gets passed a subtype-relation-closed set. I know sometimes our code excludes subtypes like this but it breaks principles of OOP.
One possible solution is to make it so `Initialize` does this instead of just assigning `filter`:
```
if(!isnull(filter))
	src.filter = list()
	for(var/type in filter)
		src.filter |= typesof(type)
```
This is good when the `filter` list is potentially large but its types are not deep. On the other hand if someone passes `list(/atom)` to this it's hell, in that case you'd rather have `src.filter = list(/atom)` and then iterate over `src.filter` here and do `istype()` check for each element. So I guess the good solution would be to make a compromise. Have a boolean var to check which of the two modes it is in, this could be chosen based on the size of `filter` in `Initialize`.

The alternative soltion that is probably more than good enough is to just make this line iterate over `filter` and do `istype` checks always. Bad performance in case `filter` is large but I assume that doesn't happen often.

What are your opinions about this?

Reply:
That is probably how it should work; sleepy brain didn't write that. The list will be mostly base types, so iterating across it and doing istype is preferred.